[Blog Refresh] Product engineer vs software engineer: What's the difference? (#13956)
PostHog/posthog.com: contents/blog/product-engineer-vs-software-engineer.md

Reviewer: We should add some more links to our other content, such as newsletters related to this. For example:

https://posthog.com/newsletter/this-is-why-youre-not-shipping

https://posthog.com/newsletter/ai-coding-mistakes

  - Product engineers
---

Software companies were once dominated by two types of roles: product managers and software engineers, but this is changing. A new role is emerging: the [product engineer](/blog/what-is-a-product-engineer).

Want proof? [Look at this graph](https://www.youtube.com/watch?v=oMRgdpXLElE):

![product-engineer-trends](https://res.cloudinary.com/dmukukwp6/image/upload/v1710055416/posthog.com/contents/images/blog/product-engineer-trend.png)

> Online searches for product engineers have grown by 80% since 2021 as startups like [ourselves](https://newsletter.posthog.com/p/beyond-the-10x-engineer), [incident.io](http://incident.io/), and [Ghost](https://ghost.org/), prioritize engineers who are customer-obsessed, able to talk to users directly, and comfortable working autonomously. 

Reviewer: Is 80% still true?

Author: updating, it's now 89%!

> Online searches for product engineers have grown by 80% since 2021 as startups like [ourselves](https://newsletter.posthog.com/p/beyond-the-10x-engineer), [incident.io](http://incident.io/), and [Ghost](https://ghost.org/), prioritize engineers who are customer-obsessed, able to talk to users directly, and comfortable working autonomously. 

Reviewer: ```suggestion
Online searches for product engineers have grown by 80% since 2021 as startups like [ourselves](https://newsletter.posthog.com/p/beyond-the-10x-engineer), [incident.io](http://incident.io/), and [Ghost](https://ghost.org/), prioritize engineers who are customer-obsessed, talk to users directly, and comfortable working autonomously. 
```

#### TL;DR – Product engineers focus on building great products while software engineers focus on building reliable software. Both of them write code, but with different priorities in mind.

> **Wait, what about growth engineers?** A [growth engineer](/blog/what-is-a-growth-engineer) is a type of software engineer who focuses on projects that drive growth metrics like signups, subscriptions, and revenue. They share many characteristics with product engineers, but have a broader focus on growth across the company rather than a specific product.

## How are product and software engineers different?
Product engineers – sometimes referred to as product-minded engineers – own the product and are responsible for its successes and failures. They're empathetic towards users and care about solving their problems. They build with the user in mind, and create what users need for success.

In contrast, software engineers own the code they write. They care more about solving technical problems, optimizing systems, and writing clean code. Since they focus more on the implementation, they usually aren’t the ones making product decisions.

## What are their priorities?

> The number one priority of a product engineers prioritize is to solve users' problems.

Reviewer: Not using priority twice
```suggestion
The top priority of a product engineers is to solving users' problems.
```

They are willing to build fast, iterate, or even from scratch if needed. The tradeoffs they consider in a solution go beyond technical constraints and might include more holistic factors like the user experience, business deadlines, and product roadmap. In this way, they tend to be more pragmatic.

> Software engineers prioritize achieving the ideal technical solution.

Reviewer: ```suggestion
Software engineers prioritize the ideal technical solution.
```

They look for the best solution to the problem currently in front of them, focusing on best practices and building off prior work. This means they're usually more specialized and idealistic. Their projects usually involve more backend and system design work, and they're constantly thinking about optimizing security, reliability, and scale.

## How do they spend their time?

Product engineers still spend the majority of their time coding, but they have responsibilities outside of coding that software engineers don't.

> Product engineers also spend time talking to customers, digging into usage data, and researching the competitive landscape. They test other products, build prototypes, and brainstorm experiments to make their product better. They use this research to prioritize a roadmap that they own, and take responsibility for explaining the roadmap and gathering feedback from users and the company.

Reviewer: We can save the "also" for the next paragraph
```suggestion
Product engineers spend time talking to customers, digging into usage data, and researching the competitive landscape. They test other products, build prototypes, and brainstorm experiments to make their product better. They use this research to prioritize a roadmap that they own, and take responsibility for explaining the roadmap and gathering feedback from users and the company.
```

Product engineering also tends to be more fast-paced, scrappy, and independent than software engineering. This is especially true as AI tools are on the rise; product engineers can move even quicker by using LLMs to handle less-critical implementation details.

Software engineers code too, of course, but they also spend lots of time designing, testing, and maintaining what they build. They look for ways to optimize systems, improve scalability, and solve bugs. They are deeply focused on specific areas of technology, whether those are databases, data pipelines, backend APIs, or client-side app frameworks.

> Software engineering projects also tend to have longer timelines and cross-functional collaboration because of their specialization.

Author: I like this a lot!

## Why is demand for product engineers growing?

There's no single reason, but it's often simply because people want to ship faster. Product engineers are popular among startups and early-stage companies for this reason, even if they don’t say it. Many software engineers behave like product engineers, even if it's not their job title.

> For example, Linear, the wildly popular modern project management tool, is built [almost entirely by product engineers](https://linear.app/now/how-we-run-projects-at-linear). Engineers rotate through project leadership roles and talk directly to users through their beta program to make decisions based on user feedback rather than traditional metrics. 

Reviewer: I think this is a good example, but I'd love to see the paragraph look like this:
1. Intro to Linear (same as you have now)
2. Some specific product engineering things they do (it's a bit unclear what "engineers rotate through project leadership roles" means/relates to and the second half of sentence is quite wordy). 
3. How this helps them ship faster

## Is product engineering right for you?

Good product engineers need to:

- Be customer-obsessed.
- Enjoy talking directly to users.
- Care deeply about outcomes and impact.
- Be flexible about exact implementation.
- Thrive in fast-paced environments.

If this sounds like you, product engineering might be for you.

If, however, you want to focus on solving difficult technical problems and optimizing existing technologies and processes, then more traditional software engineering roles are a better choice.

Just remember that product engineering is more a way of working and thinking than a defined, rigid role. Product engineers should be flexible enough to go deep on technical implementation when needed. Likewise, software engineers can choose to adopt a "product engineer" mindset whenever they like.

## When should companies hire product engineers?

> Product engineers are ideal for startups and companies that want to be agile and move fast, such as B2B SaaS or consumer subscription services. They take a holistic view of the product and have the ability to move fast when needed. They're less effective for companies with large, technically complex products, like a data tool, or platforms for large enterprises, except in certain scenarios.

Author: Thanks for noting these - I didn't realize how often I did this in my writing by habit :P
